packages/grid/_modules_/grid/hooks/features/dimensions/useGridDimensions.ts

```diff
@@ -0,0 +1,209 @@
+      rootElement.appendChild(scrollDiv);
+      scrollBarSize = scrollDiv.offsetWidth - scrollDiv.clientWidth;
+      rootElement.removeChild(scrollDiv);
+    }
+
+    const viewportOuterSize: ElementSize = {
+      width: rootDimensionsRef.current.width,
+      height: rootDimensionsRef.current.height - props.headerHeight,
+    };
+
+    // TODO: Use `useCurrentPageRows`
+    const virtualRowsCount = props.pagination
+      ? Math.min(
+          visibleRowsCount - paginationState.page * paginationState.pageSize,
+          paginationState.pageSize,
+        )
+      : visibleRowsCount;
+    const pageScrollHeight = virtualRowsCount * rowHeight;
+
+    let hasScrollX = columnsTotalWidth > viewportOuterSize.width;
+    const hasScrollY =
+      pageScrollHeight + (hasScrollX ? scrollBarSize : 0) > viewportOuterSize.height;
+
+    // We recalculate the scroll x to consider the size of the y scrollbar.
+    if (hasScrollY) {
+      hasScrollX = columnsTotalWidth + scrollBarSize > viewportOuterSize.width;
+    }
+
+    const viewportInnerSize: ElementSize = {
+      height: viewportOuterSize.height - (hasScrollX ? scrollBarSize : 0),
+      width: viewportOuterSize.width - (hasScrollY ? scrollBarSize : 0),
+    };
+
+    const maximumPageSizeWithoutScrollBar = Math.floor(viewportInnerSize.height / rowHeight);
+    const rowsInViewportCount = Math.min(virtualRowsCount, maximumPageSizeWithoutScrollBar);
+
+    const newFullDimensions: GridDimensions = {
+      viewportOuterSize,
+      viewportInnerSize,
+      maximumPageSizeWithoutScrollBar,
+      rowsInViewportCount,
+      virtualRowsCount,
+      hasScrollX,
+      hasScrollY,
+    };
+
+    const prevDimensions = fullDimensionsRef.current;
+    fullDimensionsRef.current = newFullDimensions;
+
+    if (
+      newFullDimensions.maximumPageSizeWithoutScrollBar !==
+      prevDimensions?.maximumPageSizeWithoutScrollBar
+    ) {
+      apiRef.current.publishEvent(
+        GridEvents.maximumPageSizeWithoutScrollBarChange,
+        newFullDimensions.maximumPageSizeWithoutScrollBar,
+      );
+    }
+    if (newFullDimensions.viewportInnerSize.width !== prevDimensions?.viewportInnerSize.width) {
+      apiRef.current.publishEvent(
+        GridEvents.viewportInnerWidthChange,
+        newFullDimensions.viewportInnerSize.width,
+      );
+    }
+  }, [
+    apiRef,
+    props.pagination,
+    props.headerHeight,
+    props.scrollbarSize,
+    rowHeight,
+    visibleRowsCount,
+    paginationState,
+    columnsTotalWidth,
+  ]);
+
+  const resize = React.useCallback<GridDimensionsApi['resize']>(() => {
+    updateGridDimensionsRef();
+    apiRef.current.publishEvent(GridEvents.debouncedResize, rootDimensionsRef.current);
+  }, [apiRef, updateGridDimensionsRef]);
+
+  const getDimensions = React.useCallback<GridDimensionsApi['getDimensions']>(
+    () => fullDimensionsRef.current,
+    [],
+  );
+
+  const dimensionsApi: GridDimensionsApi = {
+    resize,
+    getDimensions,
```

Comment:
Can we change this to `getGridDimentions` because it may not always be clear to what element do the dimensions belong?

Reply:
https://github.com/mui-org/material-ui-x/pull/3002

Once this one is merged it will be private so no breaking change :)

Reply:
Done

Reply:
Maybe `getRootDimensions` or even `getContainerDimensions`? We never used "grid" in an API method.

Reply:
Let's go for `getRootDimensions`